Always use `Permission` objects instead of `string`

DebHugo opened the pull request:
<!-- Summarize your PR here. Keep it short and simple. -->
<!-- Explain existing problems or why this pull request is necessary -->


### Related issues & PRs
Fix #6536 
<!--
* Fixes #1
* Related to #2
-->

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
<!-- If not, you can delete this section -->
- All functions thath use Permission now just accept `Permission` instead of `Permission|string`
- Remove `DefaultPermissionNames` (use `DefaultPermissions`)
- Now `DefaultPermissions` use `RegisteryTrait` -> contains all permissions added by pocketmine
### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->
<!-- If not, you can delete this section -->

## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
<!-- If not, you can delete this section -->
BC Break (see API changes)

## Follow-up
<!-- Suggest any actions to be done before/after merging this pull request -->
<!-- For example, future changes that this PR lays the groundwork for -->

## Tests
<!--
If this PR affects gameplay or user experience in some way, it must be manually tested.
Include any screenshots or videos of manual testing here.
Any test plugin code should also be pasted here if it can't be adapted to a PHPUnit test.
-->
Test Ingame with vanilla commands but without commands added by plugins

DebHugo committed:
Remove string for permission

DebHugo committed:
Fix bug + CS

DebHugo committed:
DefaultPermissions now use RegistryTrait

DebHugo committed:
Fix bugs

DebHugo committed:
CS

DebHugo committed:
Add forgotten functions

DebHugo committed:
CS

dktapps commented on src/command/Command.php:
```diff
@@ -93,30 +94,30 @@ public function getName() : string{
+	public function setPermission(?Permission $permission) : void{
+		$permissions = [];
+		if ($permission !== null){
+			$permissions = $this->permission;
```
This is not correct, it should replace the existing permissions, not append

dktapps commented on src/command/defaults/VanillaCommand.php:
```diff
@@ -50,8 +51,8 @@ protected function fetchPermittedPlayerTarget(CommandSender $sender, ?string $ta
-			($player === $sender && $this->testPermission($sender, $selfPermission)) ||
-			($player !== $sender && $this->testPermission($sender, $otherPermission))
+			($player === $sender && ($selfPermission === null || $this->testPermission($sender, $selfPermission))) ||
+			($player !== $sender && ($otherPermission === null || $this->testPermission($sender, $otherPermission)))
```
Not clear why this is changed when testPermission() already handles it

dktapps commented on src/permission/DefaultPermissions.php:
```diff
@@ -24,95 +24,170 @@
+			$permission->addChild($candidate, false);
 		}
 		PermissionManager::getInstance()->addPermission($candidate);
+		self::_registryRegister(str_replace(".", "_", $candidate->getName()), $candidate);
```
I think we can do away with the `POCKETMINE` prefix here

dktapps commented on src/permission/Permission.php:
```diff
@@ -36,15 +36,17 @@ class Permission{
 	public function __construct(
 		private string $name,
 		Translatable|string|null $description = null,
-		private array $children = []
```
Not clear why this is removed

DebHugo replied:
It's remove because currently array signature is `array<string, bool>` but now with Permission we can't put `array<Permission, bool>` so i remove this array and use `$this->addChild()` at replacement

ShockedPlot7560 replied:
Removing the ability to define children in the constructor may prove more complex. Could we imagine keeping this and calling addChild? Or make addChild / removeChild chainable by returning them themselves.

dktapps commented on src/permission/PermissionAttachment.php:
```diff
@@ -81,29 +80,8 @@ public function clearPermissions() : void{
-			unset($this->permissions[$node]);
-		}
-		$this->recalculatePermissibles();
-	}
```
Not sure why this is removed

DebHugo replied:
These are the same reasons as above. We remove string usage for Permission so we can't set/remove multiple Permissions just we they name, we need the Permission object. It's can be replace by for-each loop by call `setPermission`/`unsetPermission`

dktapps replied:
`unsetPermissions()` isn't affected by this.

DebHugo replied:
why ? I think it's illogic to remove usage of string but not here

dktapps replied:
There's no need to remove it since there's no technical reason to.

dktapps commented on src/plugin/PluginBase.php:
```diff
@@ -174,7 +175,7 @@ private function registerYamlCommands() : void{
 			$newCmd->setAliases($aliasList);
 
-			$newCmd->setPermission($data->getPermission());
+			$newCmd->setPermission(PermissionManager::getInstance()->getPermission($data->getPermission()));
```
This introduces a security hole, `PermissionManager::getPermission()` may return `null`

DebHugo replied:
No because this permission is already registered at https://github.com/pmmp/PocketMine-MP/blob/stable/src/plugin/PluginManager.php#L195

dktapps replied:
The permission may not have been declared.

DebHugo replied:
so i need to create new Permission here ?

dktapps replied:
No, it needs to check whether the permission exists before proceeding. Previously `setPermission()` would have checked it.

dktapps reviewed (changes requested):
After seeing the changes I'm less convinced of the value of doing this. The issue is that using `Permission` objects does not guarantee that the permission in question is actually registered.

DebHugo committed:
Fix setPermission incorect work

DebHugo committed:
Remove useless checks

DebHugo committed:
Remove POCKETMINE prefix from DefaultPermissions

DebHugo committed:
Readd unsetPermissions function + CS

DebHugo committed:
Check if Permission is registered before add this to command

ShockedPlot7560 commented on src/permission/DefaultPermissions.php:
```diff
@@ -24,95 +24,172 @@
+ * @method static Permission BROADCAST_USER()
+ * @method static Permission COMMAND_BAN_IP()
+ * @method static Permission COMMAND_BAN_LIST()
+ * @method static Permission COMMAND_BAN_PLAYER()
```
This does not appear to be self-generating

ShockedPlot7560 commented on src/permission/DefaultPermissions.php:
```diff
@@ -24,95 +24,172 @@
+final class DefaultPermissions{
+	use RegistryTrait;
+
+	public const ROOT_CONSOLE = "pocketmine.group.console";
+	public const ROOT_OPERATOR = "pocketmine.group.operator";
+	public const ROOT_USER = "pocketmine.group.user";
 
 	/**
 	 * @param Permission[] $grantedBy
 	 * @param Permission[] $deniedBy
 	 */
-	public static function registerPermission(Permission $candidate, array $grantedBy = [], array $deniedBy = []) : Permission{
+	private static function registerPermission(Permission $candidate, array $grantedBy = [], array $deniedBy = []) : Permission{
 		foreach($grantedBy as $permission){
-			$permission->addChild($candidate->getName(), true);
+			$permission->addChild($candidate, true);
 		}
 		foreach($deniedBy as $permission){
-			$permission->addChild($candidate->getName(), false);
+			$permission->addChild($candidate, false);
 		}
 		PermissionManager::getInstance()->addPermission($candidate);
 
+		$name = str_replace("pocketmine.", "", $candidate->getName());
+		self::_registryRegister(str_replace(".", "_", $name), $candidate);
+
 		return PermissionManager::getInstance()->getPermission($candidate->getName());
 	}
 
+	protected static function setup() : void{
+		self::registerCorePermissions();
+	}
+
 	public static function registerCorePermissions() : void{
-		$consoleRoot = self::registerPermission(new Permission(self::ROOT_CONSOLE, l10n::pocketmine_permission_group_console()));
-		$operatorRoot = self::registerPermission(new Permission(self::ROOT_OPERATOR, l10n::pocketmine_permission_group_operator()), [$consoleRoot]);
-		$everyoneRoot = self::registerPermission(new Permission(self::ROOT_USER, l10n::pocketmine_permission_group_user()), [$operatorRoot]);
+		$consoleRoot = self::registerPermission(new Permission("pocketmine.group.console", l10n::pocketmine_permission_group_console()));
+		$operatorRoot = self::registerPermission(new Permission("pocketmine.group.operator", l10n::pocketmine_permission_group_operator()), [$consoleRoot]);
+		$everyoneRoot = self::registerPermission(new Permission("pocketmine.group.user", l10n::pocketmine_permission_group_user()), [$operatorRoot]);
```
why removing the use of the const ?

dktapps replied:
Constants aren't needed because the permissions will be referenced by registry. It's no different to the likes of VanillaItems and VanillaBlocks.

ShockedPlot7560 replied:
The ROOT_* constants at the top of this class can then be deleted?

dktapps replied:
Probably yes, although I didn't look closely if they were still used anywhere

ShockedPlot7560 replied:
In any case, if I've understood the PR correctly, if you have to use const, you're doing it wrong because the aim of this PR is to remove the use of string in favour of the Permission object.
As far as I'm concerned, const should be removed, to be confirmed.

dktapps replied:
Yep

ShockedPlot7560 reviewed (changes requested):
I'm on the way of @dktapps for the real value of this. As Permission doesn't guarantee that the object itself is registered, I don't necessarily see the need to manipulate objects. After that, if you manage to find a system that permit it, why not, but I don't see any solution for the moment.

DebHugo commented:
> After that, if you manage to find a system that permit it, why not, but I don't see any solution for the moment.

I thought at 2 solutions:
- create `Permission::create` function with same parameters thath the constructor and put the constructor `private`. It's like the constructor but register automaticly the `Permission` to `PermissionManager` (Big BC-Break)
- In `Permission` constructor, registery automaticly the permission to `PermissionManager` (No BC-Brreak for this but it's less logic)

ShockedPlot7560 commented:
The best solution would be the PermissionManager returning the Permission object and you need to create it via PermissionManager to use them. But idk how this could be applied in a clean way

dktapps commented:
> The best solution would be the PermissionManager returning the Permission object and you need to create it via PermissionManager to use them. But idk how this could be applied in a clean way

This wouldn't prevent people from doing `new Permission` and not registering it

dktapps commented on src/Server.php:
```diff
@@ -1007,8 +1007,6 @@ public function __construct(
 			$this->logger->info($this->language->translate(KnownTranslationFactory::pocketmine_server_license($this->getName())));
 
-			DefaultPermissions::registerCorePermissions();
-
```
Only just caught this, but this could be problematic if plugins register permissions that conflict with the built-in ones. `RegistryTrait` will only register the permissions when the first permission is accessed. Sure, technically `SimpleCommandMap` should trigger it but I don't feel happy about relying on that.

ShockedPlot7560 replied:
Protecting the namespace `pocketmine.` for registering permission could be a solution. I don't really like it but I don't see another way to do it atm

dktapps replied:
Not sure how that would work without pre-registering the permissions somehow. Maybe `PermissionManager::make()` should do it similar to how we setup states for `RuntimeBlockStateRegistry`.

DebHugo replied:
I think plugins don't must remove/override a default permission

dktapps replied:
It's not good enough to just say "pls don't do this". It shouldn't be possible to begin with.

DebHugo replied:
Yes, it's must be add in other PR. Maybe, i can do this if it's necessary for submit this pr

dktapps replied:
It's a problem that this PR is introducing. Because of the use of `RegistryTrait`, the permissions will be lazily loaded, which could easily end up being after plugins load if someone makes a wrong refactor. Relying on lazy loading for something like this is risky.

The old code's explicit initialization made sure this wouldn't be a problem, but now it's gone.

DebHugo replied:
Ok so i need to re register `DefaultPermissions` after loadings of plugin ? it's that ?

dktapps replied:
No, it needs to be initialized explicitly here, as it was originally.

DebHugo committed:
Remove DefaultPermissions::ROOT_* constants

DebHugo commented on src/permission/PermissibleBase.php:
```diff
@@ -28,12 +28,8 @@ final class PermissibleBase implements Permissible{
-	public function __construct(array $basePermissions){
-		$this->permissibleBase = new PermissibleInternal($basePermissions);
+	public function __construct(){
+		$this->permissibleBase = new PermissibleInternal();
```
I'm not sure of this change. What's the utility of `permissibleBase` ?

dktapps replied:
It's for getting rid of cyclic references. Not related to the PR topic.

DebHugo committed:
Generate DefaultPermissions doc

DebHugo committed:
Reverse wrong change

DebHugo committed:
Merge branch 'major-next' into permission

DebHugo commented:
How do you want me to solve this problem ? I see solutions but it's horible solutions

dktapps commented:
Without breaking BC, the only way I see is to have `Permission::__construct()` call `PermissionManager::addPermission()`. I can't think of a use-case for constructing an unregistered permission anyway.

dktapps committed:
Fix explicit DefaultPermissions registration

perhaps we should have checkInit() public by default?

I'm not sure that explicitly initializing this registry is the proper solution, though...

dktapps committed:
Remove unused PHP_INT_MAX constant import

DebHugo committed:
Automatically register Permission when they are created

DebHugo committed:
Remove the ability to delete registered permissions

DebHugo committed:
Merge branch 'major-next' into permission

DebHugo committed:
Merge branch 'major-next' into permission

ShockedPlot7560 commented on src/permission/PermissionAttachment.php:
```diff
@@ -123,8 +111,8 @@ public function setPermission(Permission|string $name, bool $value) : void{
 
-	public function unsetPermission(Permission|string $name) : void{
-		$name = $name instanceof Permission ? $name->getName() : $name;
+	public function unsetPermission(Permission $permission) : void{
+		$name = $permission->getName();
 		if(isset($this->permissions[$name])){
```
Would be great to keep the ability to unset by the name.

ShockedPlot7560 commented on src/plugin/PluginBase.php:
```diff
@@ -174,7 +175,10 @@ private function registerYamlCommands() : void{
 			$newCmd->setAliases($aliasList);
 
-			$newCmd->setPermission($data->getPermission());
+			$permission = PermissionManager::getInstance()->getPermission($data->getPermission());
+			if ($permission !== null) {
+				$newCmd->setPermission($permission);
+			}
```
An exception should be raised in this case, no ?
